### DISCUSSION TITLE
dipy.align.reslice

### DISCUSSION THREAD

**Discussion by jagruti8** [dipy/dipy#3578] [1. Ideas & Feedback]
May I know if dipy.align.reslice takes care of aliasing. I need to downsample diffusion MRI images from 1.25*1.25*1.25 to 2.3*2.3*2.3 and anatomical images from 0.7*0.7*0.7 to 1.5*1.5*1.5

**Comment by grlee77** [comment]
It looks like reslice calls `scipy.ndimage.affine_transform` without any anti-aliasing prefilter. When downsizing, you could first apply smoothing with `scipy.ndimage.gaussian_filter` as in scikit-image's `resize` (see: https://github.com/scikit-image/scikit-image/blob/b7d0209c2b60edfe5ed1b578441c74daab4652d7/skimage/transform/_warps.py#L177-L178 and the lines above for how the default standard deviation for the Gaussian is chosen).  Perhaps this could be an option added to the `reslice` command in DIPY.

**Comment by jagruti8** [comment]
Hi, 
Thanks a lot. But do I need to apply this gaussian blurring even for downsampling from 1.25*1.25*1.25 to 1.5*1.5*1.5? Meaning, my question might be naive but how do I know if I need to blur before downsampling? Does it have to be for any downsampling or for downsampling beyond a certain point?

**Comment by jagruti8** [comment]
> It looks like reslice calls `scipy.ndimage.affine_transform` without any anti-aliasing prefilter. When downsizing, you could first apply smoothing with `scipy.ndimage.gaussian_filter` as in scikit-image's `resize` (see: https://github.com/scikit-image/scikit-image/blob/b7d0209c2b60edfe5ed1b578441c74daab4652d7/skimage/transform/_warps.py#L177-L178 and the lines above for how the default standard deviation for the Gaussian is chosen). Perhaps this could be an option added to the `reslice` command in DIPY.

So, will sigma be equal to (1.5/1.25 -1)/2 as per this?

**Comment by grlee77** [comment]
> So, will sigma be equal to (1.5/1.25 -1)/2 as per this?

That should be a reasonable starting point, but I do not know exactly how that specific value was chosen. The Gaussian filter is a lowpass filter, but does not entirely eliminate higher frequencies. It is basically a tradeoff where larger values would more strongly reduce aliasing, but at cost of some resolution loss.

I would guess for this relatively small amount of downsampling, you may not notice much visual difference with or without the antialiasing filter, but I would recommend using it in general.

Here is a quick example adapted from the [Wikipedia article on Aliasing](https://en.wikipedia.org/wiki/Aliasing):
![Moire_pattern_of_bricks](https://user-images.githubusercontent.com/6528957/128512377-30bf58ff-cc47-43dd-9d65-8948b776259b.jpg)

Without any antialiasing, downsampling by 3 looks like:
![antialiasing_none](https://user-images.githubusercontent.com/6528957/128512439-6015ba2b-cad6-49bb-9001-673559849180.png)

The scikit-image deafult antialiasing removes most of this aliasing, but you can see some residual effect:
![antialiasing_default](https://user-images.githubusercontent.com/6528957/128512515-a8a32f6e-beb9-4eeb-852f-e7884e7a1aa2.png)

If I filter with **double** the default sigma, the aliasing is further reduced, but resolution has been lost:
![antialiasing_excessive](https://user-images.githubusercontent.com/6528957/128512608-527cc241-0594-49d1-9449-03a30d8c06ee.png)

**Comment by jagruti8** [comment]
> > So, will sigma be equal to (1.5/1.25 -1)/2 as per this?
> 
> That should be a reasonable starting point, but I do not know exactly how that specific value was chosen. The Gaussian filter is a lowpass filter, but does not entirely eliminate higher frequencies. It is basically a tradeoff where larger values would more strongly reduce aliasing, but at cost of some resolution loss.
> 
> I would guess for this relatively small amount of downsampling, you may not notice much visual difference with or without the antialiasing filter, but I would recommend using it in general.
> 
> Here is a quick example adapted from the [Wikipedia article on Aliasing](https://en.wikipedia.org/wiki/Aliasing):
> ![Moire_pattern_of_bricks](https://user-images.githubusercontent.com/6528957/128512377-30bf58ff-cc47-43dd-9d65-8948b776259b.jpg)
> 
> Without any antialiasing, downsampling by 3 looks like:
> ![antialiasing_none](https://user-images.githubusercontent.com/6528957/128512439-6015ba2b-cad6-49bb-9001-673559849180.png)
> 
> The scikit-image deafult antialiasing removes most of this aliasing, but you can see some residual effect:
> ![antialiasing_default](https://user-images.githubusercontent.com/6528957/128512515-a8a32f6e-beb9-4eeb-852f-e7884e7a1aa2.png)
> 
> If I filter with **double** the default sigma, the aliasing is further reduced, but resolution has been lost:
> ![antialiasing_excessive](https://user-images.githubusercontent.com/6528957/128512608-527cc241-0594-49d1-9449-03a30d8c06ee.png)

Hi, Thanks a lot for the reply, this sigma was chosen from the link as provided:
https://github.com/scikit-image/scikit-image/blob/b7d0209c2b60edfe5ed1b578441c74daab4652d7/skimage/transform/_warps.py#L177-L178 
anti_aliasing_sigma : {float, tuple of floats}, optional
        Standard deviation for Gaussian filtering to avoid aliasing artifacts.
        By default, this value is chosen as (s - 1) / 2 where s is the
        down-scaling factor, where s > 1. For the up-size case, s < 1, no
        anti-aliasing is performed prior to rescaling.

factors = (np.asarray(input_shape, dtype=float) /
               np.asarray(output_shape, dtype=float))

anti_aliasing_sigma = np.maximum(0, (factors - 1) / 2)

image = ndi.gaussian_filter(image, anti_aliasing_sigma,
                                    cval=cval, mode=ndi_mode)

**Comment by grlee77** [comment]
> Thanks a lot for the reply, this sigma was chosen from the link as provided:
https://github.com/scikit-image/scikit-image/blob/b7d0209c2b60edfe5ed1b578441c74daab4652d7/skimage/transform/_warps.py#L177-L178

Thanks, I understood that. I meant that I wasn't sure what reference scikit-image got that default from or if it was just chosen empirically.

**Comment by jagruti8** [comment]
Thanks a lot for the reply. May I know which order should I use if I am downsampling the image using dipy.align.reslice?
order ranges from 0 to 5
default is 1 - Trilinear interpolation
